File: spring-user/src/main/java/com/example/springuser/config/SecurityConfig.java
package com.example.springuser.config;

import com.example.springuser.security.filter.JwtAuthenticationFilter;
import com.example.springuser.security.filter.JwtAuthorizationFilter;
import com.example.springuser.jwt.JwtUtil;
import com.example.springuser.security.UserDetailsServiceImpl;
// import org.springframework.boot.autoconfigure.security.servlet.PathRequest;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration;
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityCustomizer;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;

@Configuration
@EnableWebSecurity // Spring Security 지원을 가능하게 함
@EnableMethodSecurity(securedEnabled = true) // secured 어노테이션 활성화
public class SecurityConfig {

    private final JwtUtil jwtUtil;
    private final AuthenticationConfiguration authenticationConfiguration;
    private final UserDetailsServiceImpl userDetailsService;

    public SecurityConfig(JwtUtil jwtUtil,
                          AuthenticationConfiguration authenticationConfiguration,
                          UserDetailsServiceImpl userDetailsService) {
        this.jwtUtil = jwtUtil;
        this.authenticationConfiguration = authenticationConfiguration;
        this.userDetailsService = userDetailsService;
    }

    @Bean
    public AuthenticationManager authenticationManager(AuthenticationConfiguration configuration) throws Exception {
        return configuration.getAuthenticationManager();
    }

    @Bean
    public JwtAuthenticationFilter jwtAuthenticationFilter() throws Exception {
        JwtAuthenticationFilter filter = new JwtAuthenticationFilter(jwtUtil);
        filter.setAuthenticationManager(authenticationManager(authenticationConfiguration));
        return filter;
    }

    @Bean
    public JwtAuthorizationFilter jwtAuthorizationFilter() {
        return new JwtAuthorizationFilter(jwtUtil, userDetailsService);
    }

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {

        // CSRF 설정
        http.csrf((csrf) -> csrf.disable());

        // 기본 설정인 Session 방식은 사용하지 않고 JWT 방식을 사용하기 위한 설정
        http.sessionManagement((sessionManagement) ->
                sessionManagement.sessionCreationPolicy(SessionCreationPolicy.STATELESS)
        );

        http.authorizeHttpRequests((authorizeHttpRequests) ->
                authorizeHttpRequests
//                        .requestMatchers(PathRequest.toStaticResources().atCommonLocations()).permitAll() // resources 접근 허용 설정, 정적 리소스의 경우, 필터를 거치는 것이 더 손해이기 때문에, 필터를 거치지 않도록 설정
                        .requestMatchers("/").permitAll()
                        .requestMatchers("/api/user/**").permitAll() // '/api/user/'로 시작하는 요청 모두 접근 허가
                        .anyRequest().authenticated() // 그 외 모든 요청 인증처리
        );

        // 로그인 사용
        http.formLogin((formLogin) ->
                formLogin
                        // 로그인 View 제공 (GET /api/user/login-page)
                        .loginPage("/api/user/login-page")
                        // 로그인 처리 (POST /api/user/login)
                        .loginProcessingUrl("/api/user/login")
                        // 로그인 처리 후 성공 시 URL
                        .defaultSuccessUrl("/")
                        // 로그인 처리 후 실패 시 URL
                        .failureUrl("/api/user/login-page?error")
                        .permitAll()
        );

        // filter 추가
        http.addFilterBefore(jwtAuthorizationFilter(), JwtAuthenticationFilter.class); // 인가
        //  UsernamePasswordAuthenticationFilter 수행 전에, jwtAuthorizationFilter 수행 (인가 -> 로그인 진행이 올바른 순서이기 때문)
        http.addFilterBefore(jwtAuthenticationFilter(), UsernamePasswordAuthenticationFilter.class); // 인증

        // 접근 불가 페이지
        http.exceptionHandling((exceptionHandling) ->
                exceptionHandling
                        // "접근 불가" 페이지 URL 설정
                        .accessDeniedPage("/access-denied")
        );

        return http.build();
    }

    // 정적 리소스 Ignore
    @Bean
    public WebSecurityCustomizer webSecurityCustomizer() {
        return (web) -> web.ignoring()
                .requestMatchers("/js/**")
                .requestMatchers("/css/**")
                .requestMatchers("/images/**");
    }

    @Bean
    public BCryptPasswordEncoder bCryptPasswordEncoder() {
        return new BCryptPasswordEncoder();
    }
}

File: spring-kotlin-user/src/main/kotlin/com/example/kotlin/user/common/authority/SecurityConfig.kt
package com.example.kotlin.user.common.authority

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityCustomizer
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.crypto.factory.PasswordEncoderFactories
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.security.access.AccessDeniedException

@Configuration
@EnableWebSecurity
class SecurityConfig(
    private val jwtTokenProvider: JwtTokenProvider
) {
    /*
    * SecurityFilterChain
    * - httpBasic { it.disable() } : basic auth 끄기(HTTP Basic 인증을 비활성화)
    * - .csrf { it.disable() } : csrf 끄기, 주로 REST API에서 사용
    * - .sessionManagement { it.sessionCreationPolicy(SessionCreationPolicy.STATELESS) } : JWT를 사용하기 때문에 세션은 사용하지 않음
    * - .addFilterBefore(A, B) : B A (A B ): B 실행하기 전에 A 필터 실행하기(A가 통과하면, B 실행는 실행 안함)
    *
    * authorizeHttpRequests: 요청에 대한 권한을 설정
    * - hasRole('role1') (role1): 권한(role1)을 가지고 있는 경우
    * - hasAnyRole('role1', 'role2'): 권한들(role1, role2) 중 하나라도 가지고 있는 경우, 갯수 제한은 없음
    * - permitAll(): 권한 있든 말든 모두 접근 가능하다.
    * - denyAll() : 권한 있든 말든 모두 접근 불가능하다.
    * - anonymous(): Anonymous 사용자일 경우(인증을 하지 않은 사용자)
    * - rememberMe(): Remember-me기능으로 로그인한 사용자일 경우
    * - authenticated(): Anonymous 사용자가 아닌 경우(인증을 한 사용자)
    * - fullyAuthenticated(): Anonymous 사용자가 아니고  Remember-me 기능으로 로그인을 하지 않은 사용자일 경우
    *
    * [ 잎으로 작업해야 할 사항 ]
    *  Swagger 는 관리자(혹은 개발자)만 접근 가능하도록 권한을 수정 할 것
    *  실제 프로덕션 환경에서는 보안 요구사항에 따라 더 세밀한 접근 제어가 필요함
    *  - 특정 관리자 기능에 대한 추가적인 인증 요구
    *  - API 버전별 다른 인증 정책 적용
    *  - 역할 기반 접근 제어(RBAC) 구현
     */
    @Bean
    fun filterChain(http: HttpSecurity): SecurityFilterChain {
        http
            .httpBasic { it.disable() }
            .csrf { it.disable() }
            .sessionManagement {
                it.sessionCreationPolicy(SessionCreationPolicy.STATELESS)
            }
            .authorizeHttpRequests {
                it.requestMatchers("/api/members/**")
                    .authenticated()
                it.anyRequest().permitAll()
            }
            .exceptionHandling {
                it.authenticationEntryPoint { request, response, authException ->
                    throw AccessDeniedException("Authentication failed")
                }
            }
            .addFilterBefore(
                JwtAuthenticationFilter(jwtTokenProvider),
                UsernamePasswordAuthenticationFilter::class.java
            )
        return http.build()
    }

    @Bean
    fun webSecurityCustomizer(): WebSecurityCustomizer {
        return WebSecurityCustomizer { web ->
            web.ignoring()
                .requestMatchers("/js/**", "/css/**", "/images/**") // 정적 리소스 ignore 설정
        }
    }

    @Bean
    fun passwordEncoder(): PasswordEncoder =
        PasswordEncoderFactories.createDelegatingPasswordEncoder()
}